### PR TITLE
SLING-12478 Add potential origins of violations to the RequestDispatc…

### PR DESCRIPTION
**Pull request by RemoLiechti**
…her include API

### CODE REVIEW

File: src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java [hunk 1]
```diff
@@ -339,28 +350,91 @@ private Optional<String> checkContentTypeOverride(@Nullable String contentType)
...
+     * @return a string containing the unmatched TIMER_START messages
+     */
+    private String findUnmatchedTimerStarts() {
+        Iterator<String> messages = requestData.getRequestProgressTracker().getMessages();
+        List<String> unmatchedStarts = new ArrayList<>();
+        Stack<String> timerStack = new Stack<>();
+
+        // this regex matches TIMER_START{ followed by any characters except }, and then
+        // a closing }. The part inside the braces is captured for later use.
+        Pattern startPattern = Pattern.compile("TIMER_START\\{([^}]+)\\}");
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java:364]: Wouldn't it make sense to convert these into constants?

File: src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java [hunk 2]
```diff
@@ -339,28 +350,91 @@ private Optional<String> checkContentTypeOverride(@Nullable String contentType)
...
+        return unmatchedStarts.toString().trim();
+    }
+
+    /**
+     * Retrieves the message to log when the 'Content-Type' header is changed via an
+     * include.
+     *
      * @param currentContentType the current 'Content-Type' header
-     * @param setContentType the 'Content-Type' header that is being set
+     * @param setContentType     the 'Content-Type' header that is being set
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java:406]: Can you please remove this whitespace change?

File: src/test/java/org/apache/sling/engine/impl/SlingHttpServletResponseImplTest.java [hunk 3]
```diff
@@ -90,12 +134,14 @@ public void testContentMethods() {
...
-                .log(String.format(
-                        "ERROR: Servlet %s tried to override the 'Content-Type' header from 'null'"
-                                + " to 'text/plain'. This is a violation of the RequestDispatcher.include() contract -"
-                                + " https://jakarta.ee/specifications/servlet/4.0/apidocs/javax/servlet/requestdispatcher#include-javax.servlet.ServletRequest-javax.servlet.ServletResponse-.",
-                        ACTIVE_SERVLET_NAME));
+        ArgumentCaptor<String> logCaptor = ArgumentCaptor.forClass(String.class);
+        verify(requestProgressTracker, times(1)).log(logCaptor.capture());
+        String logMessage = logCaptor.getValue();
+        assertEquals(
+                String.format(
```
Comment (joerghoh) [inline src/test/java/org/apache/sling/engine/impl/SlingHttpServletResponseImplTest.java:141]: I fail to see how this large log message helps me to understand where and why the change the content-type was changed. I would appreciate if you could provide an example how to interpret this lengthy log message (preferably in the ticket SLING-12478).

Comment (RemoLiechti) [reply src/test/java/org/apache/sling/engine/impl/SlingHttpServletResponseImplTest.java:141]: Commented on [SLING-12478](https://issues.apache.org/jira/browse/SLING-12478) for reasoning.

**Review by rombert — CHANGES_REQUESTED** [review]
Overall LGTM but please make sure to remove the "Request Processing" part from the included scripts, it does not belong there.

**Comment by RemoLiechti** [comment]
> Overall LGTM but please make sure to remove the "Request Processing" part from the included scripts, it does not belong there.

see https://github.com/apache/sling-org-apache-sling-engine/pull/53/commits/c085e2e983d84f97d788cf24f799cb6937605be6

**Review by rombert — APPROVED** [review]
This looks good from my point of view. @joerghoh ?

File: src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java [hunk 4]
```diff
@@ -339,28 +360,87 @@ private Optional<String> checkContentTypeOverride(@Nullable String contentType)
...
+                    setContentType,
+                    unmatchedStartTimers,
+                    allMessages);
         }
         return String.format(
                 "Servlet %s tried to override the 'Content-Type' header from '%s' to '%s', however the"
                         + " %s forbids this via the %s configuration property. This is a violation of the "
                         + "RequestDispatcher.include() contract - "
-                        + "https://jakarta.ee/specifications/servlet/4.0/apidocs/javax/servlet/requestdispatcher#include-javax.servlet.ServletRequest-javax.servlet.ServletResponse-.",
+                        + "https://jakarta.ee/specifications/servlet/4.0/apidocs/javax/servlet/requestdispatcher#include-javax.servlet.ServletRequest-javax.servlet.ServletResponse-. , Including scripts: %s. All RequestProgressTracker messages: %s",
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java:436]: nitpick regarding the term ``Ìncluding scripts``. It can be understood like an action "I included these scripts". But if I understand it correctly, you rather want to express that this the stack of scripts, which have been included until the script was invoked which triggered the warning.
And for that I feel it should be more hinting to these call stack aspect, and less to the "include" aspect. I don't have a better suggestion, though :-/

**Review by joerghoh — APPROVED** [review]
LGTM.

If you don't have a good idea how to improve the mentioned nitpick below (I don't have any suggestion), feel free to merge.

File: src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java (hunk 4 above)
Comment (rombert) [reply src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java:436]: I see 
1. Keep this as-is
1. Include chain : Script A → Script B → Script C → (name of current script)
1. Include stack: Script A → Script B → Script C → (name of current script)

@joerghoh @RemoLiechti - thoughts?

Comment (RemoLiechti) [reply src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java:436]: I like option 3
Include stack: Script A → Script B → Script C

name of current script would automatically be the last one, so C in this case.

Comment (joerghoh) [reply src/main/java/org/apache/sling/engine/impl/SlingHttpServletResponseImpl.java:436]: +1 for option 3.

**Comment by rombert** [comment]
Thanks for the contribution @RemoLiechti !
